R/estimate_ascertainment.R

```diff
@@ -0,0 +1,144 @@
+#' "ascertainment_mean" (for the central estimate), and "ascertainment_low" and
+#' "ascertainment_high" for the lower and upper interval limits.
+#' @export
+#'
+#' @examples
+#' # get data pre-loaded with the package
+#' data("covid_data")
+#' df_covid_uk <- covid_data[covid_data$country == "United Kingdom", ]
+#'
+#' df_covid_uk_subset <- subset(df_covid_uk, date <= "2020-05-31")
+#'
+#' # use a severity baseline of 1.4% (0.014) taken from Verity et al. (2020)
+#' # Lancet Infectious Diseases: 10.1016/S1473-3099(20)30243-7
+#'
+#' # use onset-to-death distribution from Linton et al. (2020)
+#' # J. Clinical Medicine: 10.3390/jcm9020538
+#'
+#' estimate_ascertainment(
+#'   data = df_covid_uk,
+#'   delay_density = function(x) dlnorm(x, meanlog = 2.577, sdlog = 0.440),
+#'   type = "varying",
+#'   severity_baseline = 0.014,
+#'   burn_in = 7L,
+#'   max_date = as.Date("2020-06-30")
+#' )
+#'
+estimate_ascertainment <- function(data,
+                                   delay_density = NULL,
+                                   type = c("static", "varying"),
+                                   severity_baseline,
+                                   burn_in = 7,
+                                   smoothing_window = NULL,
+                                   max_date = NULL) {
+  # input checking
+  # expect rows more than burn in value
+  checkmate::assert_data_frame(data, min.cols = 3, min.rows = burn_in + 1)
+  # check that input `<data.frame>` has columns date, cases, and deaths
+  checkmate::assert_names(
+    colnames(data),
+    must.include = c("date", "cases", "deaths")
+  )
+  # check for any NAs among data
+  checkmate::assert_data_frame(
+    data[, c("date", "cases", "deaths")],
+    any.missing = FALSE
+  )
+
+  checkmate::assert_number(
+    severity_baseline,
+    lower = 0.0, upper = 1.0, finite = TRUE
+  )
+  checkmate::assert_int(burn_in, lower = 0)
+  checkmate::assert_date(max_date, null.ok = TRUE)
+
+  # NOTE: delay_density is checked in estimate_outcomes() if passed and not NULL
+
+  # match argument for type
+  type <- match.arg(type, several.ok = FALSE)
+
+  # switch the output based on user specified type
+  df_severity <- switch(type,
+    static = cfr_static(
+      data,
+      delay_density = delay_density
+    ),
+    varying = {
+      df_sev <- cfr_time_varying(
+        data,
+        delay_density = delay_density,
+        smoothing_window = smoothing_window,
+        burn_in = burn_in
+      )
+
+      df_sev <- df_sev[!is.na(df_sev$severity_mean), ]
+
+      # collect the severity at the last date, or the date specified by
+      # the user in `max_date`
+      if (!is.null(max_date)) {
+        df_sev <- df_sev[df_sev$date == max_date, ]
+      } else {
+        df_sev <- df_sev[df_sev$date == max(df_sev$date), ]
+      }
+
+      # subset column names
+      df_sev <- df_sev[
+        ,
+        grepl("severity", colnames(df_sev), fixed = TRUE)
+      ]
+    }
+  )
```

Comment:
This was/is also my understanding of how this function should work, but note that the the "time_varying" ascertainment does not actually return a time series but rather a single ascertainment estimate for the dataset, _assuming that estimated severity might have varied over time_. I guess the idea is to be able to account for changes in case definitions or changes in case reporting.

Seb has suggested splitting this function by `type` and it may be good to take the implementation you've sketched here for a new `estimate_ascertainment_timevarying()`

Reply:
Just to clarify that I didn't necessarily come down on splitting by `type`, just felt that it should be consistent between cfr and ascertainment calculations one way or the other.

Reply:
Closing this without accepting as explained by Adam in the main thread.